src/Components/Endpoints/src/Discovery/PageCollectionBuilder.cs

```diff
@@ -0,0 +1,153 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// Represents the list of pages in a <see cref="ComponentApplicationBuilder"/>.
+/// </summary>
+internal class PageCollectionBuilder
+{
+    private readonly List<PageComponentBuilder> _pages = new();
+
+    internal void Combine(PageCollectionBuilder pages)
+    {
+        for (var i = 0; i < pages._pages.Count; i++)
+        {
+            var pageToAdd = pages._pages[i];
+            var found = false;
+            for (var j = _pages.Count - 1; j > 0; j--)
+            {
+                var page = _pages[j];
+                if (page.Equals(pageToAdd))
+                {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found)
+            {
+                _pages.Add(pageToAdd);
+            }
+        }
+    }
+
+    internal void Exclude(PageCollectionBuilder pages)
+    {
+        for (var i = 0; i < pages._pages.Count; i++)
+        {
+            var pageToRemove = pages._pages[i];
+            for (var j = _pages.Count - 1; j > 0; j--)
+            {
+                var page = _pages[j];
+                if (page.Equals(pageToRemove))
+                {
+                    _pages.RemoveAt(j);
+                    break;
+                }
+            }
+        }
+    }
+
+    internal void RemoveFromAssembly(string name)
+    {
+        for (var i = _pages.Count - 1; i > 0; i--)
+        {
+            if (_pages[i].HasSource(name))
+            {
+                _pages.RemoveAt(i);
+            }
+        }
+    }
+
+    internal void AddFromLibraryInfo(IEnumerable<PageComponentBuilder> pages)
+    {
+        _pages.AddRange(pages);
+    }
+
+    internal PageComponentInfo[] ToPageCollection()
+    {
+        if (_pages.Count == 0)
+        {
+            return Array.Empty<PageComponentInfo>();
+        }
+
+        var list = new List<PageComponentInfo>();
+        // Reuse a buffer for computing the metadata
+        using var buffer = new MetadataBuffer(256);
+        for (var i = 0; i < _pages.Count; i++)
+        {
+            var page = _pages[i];
+            foreach (var route in page.RouteTemplates!)
+            {
+                list.Add(new PageComponentInfo(route, page.PageType!, route, ResolveMetadata(page.PageType!, buffer)));
+            }
+        }
+
+        return list.ToArray();
+    }
+
+    private static IReadOnlyList<object> ResolveMetadata(Type componentType, MetadataBuffer buffer)
```

Comment:
Should `buffer` be passed by ref here? I was unclear on whether `ref struct` is automatically passed by ref but apparently it is not: https://stackoverflow.com/questions/75685686/why-passing-ref-struct-to-methods-actually-passes-a-copy-of-a-ref-struct.

I'm asking because the code appears to assume it's necessary to call `Reset` but if it's being passed by value then it would already be in a clean state anyway.

Apologies if I'm misreading this code - it totally might be exactly right already.

Reply:
Simplified this